walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,111 @@
+    }
+
+    // This method doesn't seem efficient. I'd appreciate suggestions
+    // TODO: Refactor for efficiency
+    private ArrayList<TimeRange> checkTimes(Collection<Event> events, 
+        MeetingRequest request, boolean optionalCheck){
+
+        ArrayList<TimeRange> bookedTimes = new ArrayList<TimeRange>();
+        for(Event e : events){
+            //Adds the time slot to booked times if it pertains to request people
+            Set<String> people = e.getAttendees();
+            for(String person : people){
+                if(request.getAttendees().contains(person) || 
+                    (request.getOptionalAttendees().contains(person) && optionalCheck)){
+                    
+                    boolean overlap = false;
+                    for(int ii = 0; ii < bookedTimes.size(); ii++){
+                        TimeRange bt = bookedTimes.get(ii);
+                        if(bt.overlaps(e.getWhen()) || bt.end() == e.getWhen().start() || 
+                            bt.start() == e.getWhen().end()){
+                            overlap = true;
+                            TimeRange newTime = 
+                                TimeRange.fromStartEnd(Math.min(e.getWhen().start(), bt.start()),
+                                Math.max(e.getWhen().end(), bt.end()), false);
+                            bookedTimes.set(ii, newTime);
+                        }
+                    }
+                    if(!overlap){
+                        TimeRange newTime = e.getWhen();
+                        TimeRange temp;
+                        // Sorts meeting by start time
+                        // Can't have overlap
+                        for(int ii = 0; ii < bookedTimes.size(); ii++){
+                            TimeRange bt = bookedTimes.get(ii);
+                            if(newTime.start() < bt.start()){
+                                temp = bt;
+                                bookedTimes.set(ii, newTime);
+                                newTime = temp;
+                            }
+                        }
+                        bookedTimes.add(newTime);
+                    }
+                    break;
+                }
+            }
+        }
+        return bookedTimes;
+    }
+
+    //Expects sorted Collection
+    public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
+        ArrayList<TimeRange> bookedTimes = new ArrayList<TimeRange>();
+        Collection<TimeRange> possibleTimes;
+        
+        //Checks to ensure valid MeetingRequest duration
+        if(request.getDuration() > TimeRange.WHOLE_DAY.duration())
+            return new ArrayList<TimeRange>();
+
+        //Checks for mandatory and optional attendees
+        bookedTimes = checkTimes(events, request, true);
+
+        if(bookedTimes.isEmpty()){
+            possibleTimes = new ArrayList<TimeRange>();
+            possibleTimes.add(TimeRange.WHOLE_DAY);
+            return possibleTimes;
+        }
```

Comment:
Is this part necessary? Would `invert(bookedTimes, request.getDuration())` properly handle the case where bookedTimes is empty?.